File: android-native/app/src/main/kotlin/com/wordwheel/game/ui/LetterWheel.kt
package com.wordwheel.game.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.*
import androidx.compose.runtime.*
import androidx.compose.runtime.snapshots.SnapshotStateList
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.text.TextMeasurer
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.toSize
import com.wordwheel.game.audio.LocalSoundManager
import com.wordwheel.game.audio.Sfx
import com.wordwheel.game.theme.GameColors
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Renders the letter wheel. [selection] is passed as a [SnapshotStateList] so
 * gesture callbacks mutate it directly — avoiding stale-closure bugs that
 * would occur with an immutable [List] + callback pattern (the pointerInput
 * block is only re-launched when [tiles] changes, so any local mutable copy
 * would go stale the moment the parent clears the selection).
 */
@Composable
fun LetterWheel(
    tiles: List<Char>,
    selection: SnapshotStateList<Int>,
    onSubmit: () -> Unit,
    onShuffle: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var wheelSize by remember { mutableStateOf(Size.Zero) }
    val textMeasurer = rememberTextMeasurer()
    val sound = LocalSoundManager.current

    Box(
        modifier = modifier
            .aspectRatio(1f)
            .onSizeChanged { wheelSize = it.toSize() }
            .pointerInput(tiles) {
                // `lastPos` tracks the previous drag sample so we can
                // interpolate a fast slide. Without interpolation a quick
                // swipe across the wheel skips tiles whose centers fall
                // between consecutive drag events.
                var lastPos: Offset? = null
                detectDragGestures(
                    onDragStart = { pos ->
                        val hit = hitTest(wheelSize, tiles.size, pos)
                        if (hit >= 0) {
                            selection.clear()
                            selection.add(hit)
                            sound?.play(Sfx.Select)
                        }
                        lastPos = pos
                    },
                    onDrag = { change, _ ->
                        val from = lastPos ?: change.position
                        val to = change.position
                        // Sample several points along the segment so a
                        // fast swipe still registers each tile it crosses.
                        // Half a tile-radius per step is small enough that
                        // we can't skip a tile, big enough to keep cost low.
                        val tileRpx = wheelSize.width.coerceAtLeast(1f) *
                            0.5f * 0.95f * 0.17f
                        val stepLen = tileRpx.coerceAtLeast(1f) * 0.5f
                        val steps = (distance(from, to) / stepLen)
                            .toInt()
                            .coerceIn(1, 12)
                        for (s in 1..steps) {
                            val t = s.toFloat() / steps
                            val p = Offset(
                                from.x + (to.x - from.x) * t,
                                from.y + (to.y - from.y) * t,
                            )
                            val hit = hitTest(wheelSize, tiles.size, p)
                            if (hit < 0 || hit == selection.lastOrNull()) continue
                            // Backtrack: dragging back over the previous
                            // tile shortens the word — feels natural and
                            // matches Wordscapes / Words of Wonders.
                            if (selection.size >= 2 &&
                                hit == selection[selection.size - 2]) {
                                selection.removeAt(selection.size - 1)
                                sound?.play(Sfx.Select)
                            } else if (!selection.contains(hit)) {
                                selection.add(hit)
                                sound?.play(Sfx.Select)
                            }
                        }
                        lastPos = to
                    },
                    onDragEnd = { lastPos = null; onSubmit() },
                    onDragCancel = { lastPos = null; onSubmit() },
                )
            }
            .pointerInput(tiles) {
                detectTapGestures(
                    onTap = { pos ->
                        // Center tap → shuffle
                        val center = Offset(wheelSize.width / 2f, wheelSize.height / 2f)
                        val shuffleR = wheelSize.width * 0.09f
                        if (distance(center, pos) <= shuffleR) {
                            sound?.play(Sfx.Shuffle)
                            onShuffle()
                            return@detectTapGestures
                        }
                        val hit = hitTest(wheelSize, tiles.size, pos)
                        if (hit >= 0 && !selection.contains(hit)) {
                            selection.add(hit)
                            sound?.play(Sfx.Select)
                        }
                    }
                )
            }
    ) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val center = Offset(size.width / 2f, size.height / 2f)
            val radius = minOf(size.width, size.height) / 2f * 0.95f
            val tileOrbit = radius * 0.6f
            val tileR = radius * 0.17f

            drawCircle(color = GameColors.WheelBg, radius = radius, center = center)
            drawCircle(
                color = Color(0x78FFFFFF),
                radius = radius,
                center = center,
                style = Stroke(width = 2.5f),
            )

            val positions = List(tiles.size) { i ->
                val angle = i.toFloat() / tiles.size * 2f * PI.toFloat() - PI.toFloat() / 2f
                Offset(
                    center.x + cos(angle) * tileOrbit,
                    center.y + sin(angle) * tileOrbit,
                )
            }

            // Selection lines
            if (selection.size >= 2) {
                for (i in 0 until selection.size - 1) {
                    val ai = selection[i]
                    val bi = selection[i + 1]
                    if (ai in positions.indices && bi in positions.indices) {
                        drawLine(
                            color = GameColors.LineColor,
                            start = positions[ai],
                            end = positions[bi],
                            strokeWidth = 8f,
                        )
                    }
                }
            }

            // Tiles
            for (i in tiles.indices) {
                val pos = positions[i]
                val isSelected = selection.contains(i)
                if (isSelected) {
                    drawCircle(
                        color = GameColors.TileSelectedBg,
                        radius = tileR,
                        center = pos,
                    )
                }
                drawTileLetter(
                    measurer = textMeasurer,
                    letter = tiles[i].toString(),
                    center = pos,
                    sizePx = tileR * 1.1f,
                    color = if (isSelected) Color.White else GameColors.LetterColor,
                )
            }

            // Shuffle icon
            drawTileLetter(
                measurer = textMeasurer,
                letter = "\u21C4",
                center = center,
                sizePx = radius * 0.26f,
                color = GameColors.ShuffleIcon,
            )
        }
    }
}

private fun distance(a: Offset, b: Offset): Float {
    val dx = a.x - b.x
    val dy = a.y - b.y
    return sqrt(dx * dx + dy * dy)
}

private fun tilePositions(size: Size, n: Int): List<Offset> {
    if (size.width == 0f || n == 0) return emptyList()
    val center = Offset(size.width / 2f, size.height / 2f)
    val radius = minOf(size.width, size.height) / 2f * 0.95f
    val tileOrbit = radius * 0.6f
    return List(n) { i ->
        val angle = i.toFloat() / n * 2f * PI.toFloat() - PI.toFloat() / 2f
        Offset(
            center.x + cos(angle) * tileOrbit,
            center.y + sin(angle) * tileOrbit,
        )
    }
}

private fun hitTest(size: Size, n: Int, pos: Offset): Int {
    if (size.width == 0f) return -1
    val radius = minOf(size.width, size.height) / 2f * 0.95f
    val tileR = radius * 0.17f
    val positions = tilePositions(size, n)
    // Generous hit padding (~40% beyond the visible tile) keeps drags
    // forgiving when fingers slide quickly — the previous +12px constant
    // was tight enough that fast swipes regularly missed tiles.
    val hitR = tileR * 1.45f
    return positions.indexOfFirst { distance(it, pos) <= hitR }
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawTileLetter(
    measurer: TextMeasurer,
    letter: String,
    center: Offset,
    sizePx: Float,
    color: Color,
) {
    val style = TextStyle(
        color = color,
        fontSize = sizePx.toSp(),
        fontWeight = FontWeight.Bold,
    )
    val layout = measurer.measure(text = letter, style = style)
    drawText(
        textLayoutResult = layout,
        topLeft = Offset(
            center.x - layout.size.width / 2f,
            center.y - layout.size.height / 2f,
        ),
    )
}

File: ios-native/WordWheel/Views/LetterWheel.swift
import SwiftUI

/// The circular letter wheel. Drag across tiles to build a word; tapping the
/// center icon shuffles the tiles.
///
/// Mirrors `ui/LetterWheel.kt` on the Android side — Canvas for drawing,
/// gesture handlers for selection, and the same geometry constants so the
/// visual output is identical.
struct LetterWheel: View {
    let tiles: [Character]
    @Binding var selection: [Int]
    let onSubmit: () -> Void
    let onShuffle: () -> Void

    @EnvironmentObject private var sound: SoundManager
    @State private var wheelSize: CGSize = .zero
    /// Last drag location captured by the gesture handler — used to
    /// interpolate intermediate hit-tests so a fast slide doesn't skip
    /// past tiles whose centers fall between two consecutive events.
    @State private var lastDragPos: CGPoint? = nil

    var body: some View {
        GeometryReader { geo in
            let size = geo.size
            let center = CGPoint(x: size.width / 2, y: size.height / 2)
            let radius = min(size.width, size.height) / 2 * 0.99
            // tileOrbit + tileR = 0.94 → tile outer edge sits 5 % inside
            // the visible disc border. The previous 0.62 + 0.20 = 0.82
            // left a 17 % "moat" of empty white space ("jarak antara
            // border dan alphabetnya terlalu jauh" in user feedback).
            let tileOrbit = radius * 0.72
            let tileR = radius * 0.22
            let shuffleR = size.width * 0.09

            let positions: [CGPoint] = (0..<tiles.count).map { i in
                let angle = Double(i) / Double(max(tiles.count, 1)) * 2 * .pi - .pi / 2
                return CGPoint(
                    x: center.x + CGFloat(cos(angle)) * tileOrbit,
                    y: center.y + CGFloat(sin(angle)) * tileOrbit,
                )
            }

            ZStack {
                // Wheel background disc
                Circle()
                    .fill(GameColors.wheelBg)
                    .overlay(Circle().stroke(Color.white.opacity(0.47), lineWidth: 2.5))

                // Selection lines
                if selection.count >= 2 {
                    Path { p in
                        let pts = selection.compactMap { positions.indices.contains($0) ? positions[$0] : nil }
                        guard let first = pts.first else { return }
                        p.move(to: first)
                        for pt in pts.dropFirst() { p.addLine(to: pt) }
                    }
                    .stroke(GameColors.lineColor, style: StrokeStyle(lineWidth: 8, lineCap: .round, lineJoin: .round))
                }

                // Tiles
                ForEach(Array(tiles.enumerated()), id: \.offset) { i, ch in
                    let isSelected = selection.contains(i)
                    ZStack {
                        if isSelected {
                            Circle()
                                .fill(GameColors.tileSelectedBg)
                                .frame(width: tileR * 2, height: tileR * 2)
                        }
                        Text(String(ch))
                            .font(.system(size: tileR * 1.1, weight: .bold))
                            .foregroundColor(isSelected ? .white : GameColors.letterColor)
                    }
                    .position(positions.indices.contains(i) ? positions[i] : center)
                }

                // Center shuffle icon
                Text("\u{21C4}")
                    .font(.system(size: radius * 0.26, weight: .bold))
                    .foregroundColor(GameColors.shuffleIcon)
                    .position(center)
            }
            .contentShape(Rectangle())
            .simultaneousGesture(
                DragGesture(minimumDistance: 2)
                    .onChanged { value in
                        // First hit starts the selection.
                        if selection.isEmpty {
                            if let h = hitTest(size: size, positions: positions,
                                               tileR: tileR, point: value.location) {
                                selection = [h]
                                sound.play(.select)
                            }
                            lastDragPos = value.location
                            return
                        }
                        // Interpolate between the previous sample and the
                        // current one. Without this, fast swipes skip
                        // tiles whose centers fall between two events.
                        let from = lastDragPos ?? value.location
                        let to = value.location
                        let dx = to.x - from.x
                        let dy = to.y - from.y
                        let dist = sqrt(dx * dx + dy * dy)
                        // Half a tile-radius per step keeps us under the
                        // tile spacing so we can't skip a tile.
                        let stepLen = max(tileR * 0.5, 1)
                        let steps = max(1, min(12, Int(dist / stepLen)))
                        for s in 1...steps {
                            let t = CGFloat(s) / CGFloat(steps)
                            let p = CGPoint(x: from.x + dx * t, y: from.y + dy * t)
                            guard let h = hitTest(size: size, positions: positions,
                                                  tileR: tileR, point: p) else { continue }
                            if h == selection.last { continue }
                            // Backtrack: dragging back over the previous
                            // tile shortens the word — matches the feel
                            // of Wordscapes / Words of Wonders.
                            if selection.count >= 2,
                               h == selection[selection.count - 2] {
                                selection.removeLast()
                                sound.play(.select)
                            } else if !selection.contains(h) {
                                selection.append(h)
                                sound.play(.select)
                            }
                        }
                        lastDragPos = to
                    }
                    .onEnded { _ in
                        lastDragPos = nil
                        onSubmit()
                    }
            )
            .onTapGesture { location in
                // Center → shuffle
                let dx = location.x - center.x
                let dy = location.y - center.y
                if sqrt(dx * dx + dy * dy) <= shuffleR {
                    sound.play(.shuffle)
                    onShuffle()
                    return
                }
                if let h = hitTest(size: size, positions: positions, tileR: tileR, point: location),
                   !selection.contains(h) {
                    selection.append(h)
                    sound.play(.select)
                }
            }
            .onAppear { wheelSize = size }
            .onChange(of: size.width) { _ in wheelSize = size }
        }
        .aspectRatio(1, contentMode: .fit)
    }

    private func hitTest(size: CGSize, positions: [CGPoint], tileR: CGFloat, point: CGPoint) -> Int? {
        // Generous hit padding (~45% beyond the visible tile) keeps the
        // gesture forgiving when fingers slide quickly. The previous
        // +12pt constant was tight enough that fast swipes regularly
        // missed tiles between samples.
        let hitR = tileR * 1.45
        for (i, p) in positions.enumerated() {
            let dx = p.x - point.x
            let dy = p.y - point.y
            if sqrt(dx * dx + dy * dy) <= hitR { return i }
        }
        return nil
    }
}
